gk/main.c

```diff
@@ -968,12 +972,241 @@ gk_setup_rss(struct gk_config *gk_conf)
 	return ret;
 }
 
+static inline unsigned short
+icmp_cksum(void *buf, int size)
+{
+	unsigned short *buffer = buf;
+	unsigned long cksum = 0;
+
+	while(size > 1) {
+		cksum += *buffer++;
+		size -= sizeof(unsigned short);
+	}
+
+	if(size)
+		cksum += *(unsigned char*)buffer;
+
+	cksum = (cksum >> 16) + (cksum & 0xffff);
+	cksum += (cksum >>16);
+
+	return (unsigned short)(~cksum);
+}
+
+static void
+xmit_icmp(struct ether_hdr *eth_hdr, struct ipv4_hdr *ipv4_hdr,
+	uint8_t port, uint16_t tx_queue, uint16_t *num_pkts,
+	struct rte_mbuf **icmp_bufs, uint16_t max_num_pkts_burst,
+	struct gk_config *gk_conf)
+{
+	uint16_t i, num_tx_succ;
+	struct ether_hdr *icmp_eth;
+	struct ipv4_hdr *icmp_ipv4;
+	struct icmp_hdr *icmph;
+
+	struct rte_mbuf *icmp_pkt = rte_pktmbuf_alloc(
+		gk_conf->net->gatekeeper_pktmbuf_pool[
+		rte_lcore_to_socket_id(rte_lcore_id())]);
+	if (icmp_pkt == NULL) {
+		RTE_LOG(ERR, MEMPOOL,
+			"gk: failed to allocate an ICMP packet!");
+		return;
+	}
+
+	icmp_eth = (struct ether_hdr *)rte_pktmbuf_append(icmp_pkt,
+		sizeof(struct ether_hdr) + sizeof(struct ipv4_hdr) +
+		sizeof(struct icmp_hdr));
+	if (icmp_eth == NULL) {
+		RTE_LOG(ERR, MBUF,
+			"Not enough headroom space in the first segment!\n");
+		return;
+	}
+
+	ether_addr_copy(&eth_hdr->s_addr, &icmp_eth->d_addr);
+	ether_addr_copy(&eth_hdr->d_addr, &icmp_eth->s_addr);
+	icmp_eth->ether_type = eth_hdr->ether_type;
+
+	icmp_ipv4 = (struct ipv4_hdr *)&icmp_eth[1];
+	icmp_ipv4->version_ihl = IP_VHL_DEF;
+	icmp_ipv4->type_of_service = 0;
+	icmp_ipv4->packet_id = 0;
+	icmp_ipv4->fragment_offset = IP_DN_FRAGMENT_FLAG;
+	icmp_ipv4->time_to_live = IP_DEFTTL;
+	icmp_ipv4->next_proto_id = IPPROTO_ICMP;
+	icmp_ipv4->src_addr = ipv4_hdr->dst_addr;
+	icmp_ipv4->dst_addr = ipv4_hdr->src_addr;
+	icmp_ipv4->total_length = rte_cpu_to_be_16(icmp_pkt->data_len);
+	/*
+	 * The IP header checksum filed must be set to 0
+	 * in order to offload the checksum calculation.
+	 */
+	icmp_ipv4->hdr_checksum = 0;
+	icmp_pkt->l2_len = sizeof(struct ether_hdr);
+	icmp_pkt->l3_len = sizeof(struct ipv4_hdr);
+	icmp_pkt->ol_flags |= PKT_TX_IPV4 | PKT_TX_IP_CKSUM;
+
+	icmph = (struct icmp_hdr *)&icmp_ipv4[1];
+	icmph->icmp_type = ICMP_TIME_EXCEEDED;
+	icmph->icmp_code = ICMP_EXC_TTL;
+	icmph->icmp_cksum = 0;
+	icmph->icmp_ident = 0;
+	icmph->icmp_seq_nb = 0;
+	icmph->icmp_cksum = icmp_cksum(icmph, sizeof(*icmph));
+
+	icmp_bufs[*num_pkts] = icmp_pkt;
+	(*num_pkts)++;
+
+	if (*num_pkts < max_num_pkts_burst)
```

Comment:
If `*num_pkts` doesn't become greater or equal to `max_num_pkts_burst`, the ICMP packets are never sent out. This may cause unexpectedly long delays in the replies.

Reply:
I think the easiest thing to do is to send out all ICMP packets available after processing a batch of packets.

Reply:
That makes sense.